source/mir/cv/imgproc/imfilter.d

```diff
@@ -0,0 +1,459 @@
+enum BORDER_SYMMETRIC = 2;
+
+// Error codes that can occur in image filtering calls.
+enum ERROR_INVALID_BORDER_METHOD = 1;
+
+/++
+    Separable image filtering.
+
+    Params:
+        inputbuf        = Input image. Size of 'rows' * 'cols'.
+        hmask           = Horizontal mask. Size of 'hsize'.
+        vmask           = Vertical mask. Size of 'vsize'.
+        outputbuf       = Output image. Size of 'rows' * 'cols'.
+        rows            = Row size (height) of input/output images.
+        cols            = Column size (width) of input/output images.
+        hsize           = Size of horizontal kernel.
+        vsize           = Size of vertical kernel.
+        border_handling = Border handling scheme used to process borders of the image.
+
+    Returns:
+        0 if successful, ERROR_INVALID_BORDER_METHOD if invalid value is passed for border_handling parameter.
++/
+pragma(inline, false)
+int separable_imfilter
+(
+    float* inputbuf,
+    float* hmask,
+    float* vmask,
+    float* outputbuf,
+    size_t rows,
+    size_t cols,
+    size_t hsize,
+    size_t vsize,
+    int border_handling = BORDER_REPLICATE
+)
+{
+    return separable_imfilter_impl
+        (inputbuf, hmask, vmask, outputbuf, rows, cols, hsize, vsize, border_handling);
+}
+
+package(mir.cv):
+
+template Horizontal_kernel_func(alias InstructionSet)
+{
+    alias Horizontal_kernel_func = void function(InstructionSet.Scalar*, // input pointer.
+        InstructionSet.Vector*, // kernel mask pointer.
+        InstructionSet.Scalar*, // output pointer.
+        size_t); // size of the kernel mask.
+}
+
+template Vertical_kernel_func(alias InstructionSet)
+{
+    alias Vertical_kernel_func = void function(InstructionSet.Scalar*, // input pointer.
+        InstructionSet.Vector*, // kernel mask pointer.
+        InstructionSet.Scalar*, // output pointer.
+        size_t, // size of the kernel mask.
+        size_t); // rowstride to read another row element.
+}
+
+pragma(inline, true)
+auto max(T)(T t1, T t2) {
+    return (t1 > t2) ? t1 : t2;
+}
+
+int separable_imfilter_impl(T)
+(
+    T* inputbuf,
+    T* hmask,
+    T* vmask,
+    T* outputbuf,
+    size_t rows,
+    size_t cols,
+    size_t hsize,
+    size_t vsize,
+    int border_handling = BORDER_REPLICATE
+)
+{
+    auto tempbuf = cast(T[])alignedAllocate(rows * cols * T.sizeof, 16); // temporary buffer, used to store horizontal filtering
+
+    scope (exit)
+    {
+        deallocate(tempbuf);
+    }
+
+    auto input = inputbuf.sliced(rows, cols).universal;
+    auto temp = tempbuf.sliced(rows, cols).universal;
+    auto output = outputbuf.sliced(rows, cols).universal;
+    auto hm = hmask.sliced(hsize).universal;
+    auto vm = vmask.sliced(vsize).universal;
+
+    if (avx) // add version (compile time flag if avx should be supported)
+    {
+        inner_filtering_impl!(AVX!T)(input, temp, hm, vm, output,
+            &apply_horizontal_kernel_simd!(AVX!T), &apply_vertical_kernel_simd!(AVX!T));
```

Comment:
`apply_vertical_kernel_simd` is marked `pragma(inline, true)`. If you want `apply_vertical_kernel_simd` to be inlined it is more intuitive to pass it as template param